### ISSUE TITLE
Proposal: Behaviors

### ISSUE THREAD

**Issue by matanlurey** [angular/angular#2292]
Story
===

Consider the following scenario:

I want to listen to a custom event that I've defined and want to re-use across my entire application, but one that does *not* exist natively in the DOM, or one I want to massage, throttle, etc. For example, `hover`.

```dart
abstract class HoverHandler {
  Stream get onHover;
}

@Directive(
  selector: '[hover]',
  hostInjector: const [
    const Binding(HoverHandler, toAlias: HoverDirective)
  ]
)
class HoverDirective implements HoverHandler  {
  final Element _domElement;

  HoverDirective(ElementRef elementRef) : _domElement = elementRef.domElement;

  @override
  Stream get onHover {
    // In reality, do a whole bunch of black magic here, but let's dumb it down for now.
    return _domElement.onMouseMove;
  }
}
```

Then, I want to use the `Hover` behavior in my `Link` component:

```dart
@Component(
  selector: 'fancy-link'
)
class FancyLinkComponent {
  FancyLinkComponent(HoverHandler hoverHandler) {
    hoverHandler.onHover.listen((_) {
      // Do something, maybe show a tooltip, change colors dynamically, etc.
    });
  }
}
```

This works, but only if I remember to put `hover` on all of my elements:

```html
<fancy-link hover></fancy-link>
```

Proposal
---

Instead, what if I could automatically add directives/decorators? For example:

```dart
@Component(
  selector: 'fancy-link',
  behaviors: const [HoverHandler]
)
class FancyLinkComponent { ... }
```

For hover, this may seem trivial, but imagine something similar to standard ARIA access:

```dart
abstract class AriaEventsHandler {
  Stream get onPress;       // E.g., mouse click, tap, space down, enter key down
  Stream get onRelease;   // E.g., mouse up, tap up, space up, enter key up
  Stream get onNext;         // E.g., keyboard right in horizontal lists, keyboard down in vertical
  Stream get onPrevious;  // Opposite of above
}

@Component(
  selector: 'fancy-select',
  behaviors: const [AriaEventsHandler]
)
class SelectComponent {
  SelectComponent(AriaEventsHandler ariaHandler) {
    ariaHandler.onPress.listen((_) { ... });
  }
}
```

Alternatives
---

Other alternatives, which require additional boilerplate, but work today:

- Manually adding decorators everywhere in the DOM
- Creating wrapper components, e.g.

```dart
@View(
  template: r'''
    <fancy-select aria></fancy-select>
  '''
)
class AriaSelectComponent {}
```

@jelbourn, @mhevery

**Comment by jelbourn** [comment]
The existing `EventManagerPlugin` sounds like it's pretty close to what you're looking for. Check out The Hammer.js plugin in the source for an example.

**Comment by mhevery** [comment]
What you are asking for is to have `<fancy-link>` add `hover` attribute and then have the compiler pick it up in the compilation. This creates a bit of magic at a distance, which I don't think is worth the complexity.

In the above case you can just have `HoverService` which you can be inject into `fancy-link` which should be able to pragmatically add the hoover behavior to the elment.

WORKING AS INTENDED
